charms/istio-pilot/lib/charms/istio_pilot/v0/ingress.py

```diff
@@ -0,0 +1,268 @@
+    """Event triggered when something went wrong with the ingress relation."""
+
+
+class IngressRemovedEvent(EventBase):
+    """Event triggered when the ingress relation is removed."""
+
+
+class IngressRequirerEvents(ObjectEvents):
+    available = EventSource(IngressProviderAvailableEvent)
+    ready = EventSource(IngressReadyEvent)
+    failed = EventSource(IngressFailedEvent)
+    removed = EventSource(IngressRemovedEvent)
+
+
+class IngressRequirer(Object):
+    on = IngressRequirerEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        *,
+        port: int = None,
+        service: str = None,
+        prefix: str = None,
+        rewrite: str = None,
+        namespace: str = None,
+        per_unit_routes: bool = False,
+    ):
+        """Constructor for IngressRequirer.
+
+        The request args can be used to specify the ingress properties when the
+        instance is created. If any are set, at least `port` is required, and
+        they will be sent to the ingress provider as soon as it is available.
+        All request args must be given as keyword args.
+
+        Args:
+            charm: the charm that is instantiating the library.
+            relation_name: the name of the relation endpoint to bind to
+                (defaults to "ingress"; relation must be of interface type
+                "ingress" and have "limit: 1")
+        Request Args:
+            service: the name of the target K8s service to route to; defaults to the
+                charm's automatically created service (i.e., the application name)
+            port: the port of the service (required)
+            prefix: the path used to match this service for requests to the gateway;
+                must not conflict with other services; defaults to f"/{service}/"
+            rewrite: the path on the target service to map the request to; defaults
+                to "/"
+            namespace: the namespace the service is in; default to the current model
+            per_unit_routes: whether or not to create URLs which map to specific units;
+                the URLs will have their own prefix of f"{prefix}-unit-{unit_num}" (with
+                tailing slashes handled appropriately)
+        """
+        super().__init__(charm, f"ingress-requirer-{relation_name}")
+        self.charm = charm
+        self.relation_name = relation_name
+
+        self._validate_relation_meta()
+
+        self.status = self._get_status()
+
+        self._request_args = {
+            "port": port,
+            "service": service,
+            "prefix": prefix,
+            "rewrite": rewrite,
+            "namespace": namespace,
+            "per_unit_routes": per_unit_routes,
+        }
+        if any(self._request_args.values()) and not self._request_args["port"]:
+            raise TypeError("Missing required argument: 'port'")
+
+        if charm.unit.is_leader():
```

Comment:
This breaks access to per-unit routing info to non-leader units

Reply:
Yes, but it's currently required to work around https://github.com/canonical/serialized-data-interface/issues/27. Doesn't the new prometheus interface also use app-level relation data, in which case only the leader need be involved anyway?

Reply:
Nope. For Prometheus, each unit is on its own.